### PR TITLE
Wip dang zipper 8

### PR DESCRIPTION
**Pull request by dang**
Zipper 8 - the zippening!

This has one major feature: it splits the RGWRadosStore into a separate set of files from RGWStore and company.  It also has a lot of cleanup, removing API-breaking calls to getRados() and adding missing APIs.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dang** [comment]
https://pulpito.ceph.com/dang-2020-08-25_14:52:18-rgw-wip-dang-zipper-8-distro-basic-smithi/

That's a lot of failures, but most of them are ragweed.  I've verified that every other failure happened before this patch, so it's "clean" as far as that goes.

File: src/rgw/librgw.cc [hunk 1]
```diff
@@ -227,8 +227,8 @@ namespace rgw {
     rgw_env.set("HTTP_HOST", "");
 
     /* XXX and -then- bloat up req_state with string copies from it */
-    struct req_state rstate(req->cct, &rgw_env, req->get_user(), req->id);
-    struct req_state *s = &rstate;
+    struct req_state rstate(req->cct, &rgw_env, req->id);
+    struct req_state *s = req->get_state();
```
Comment (cbodley) [inline src/rgw/librgw.cc:231]: can `rstate` be safely removed now?

File: src/rgw/rgw_sal.h [hunk 2]
```diff
@@ -85,6 +73,10 @@ class RGWStore : public DoutPrefixProvider {
     virtual bool is_meta_master() = 0;
     virtual int forward_request_to_master(RGWUser* user, obj_version *objv,
 				  bufferlist& in_data, JSONParser *jp, req_info& info) = 0;
+    virtual int defer_gc(RGWObjectCtx *rctx, RGWBucket* bucket, RGWObject* obj,
+			 optional_yield y) = 0;
+    virtual const RGWZoneGroup& get_zonegroup() = 0;
+    virtual int get_zonegroup(const string& id, RGWZoneGroup& zonegroup) = 0;
```
Comment (cbodley) [inline src/rgw/rgw_sal.h:79]: i'm not sure how much multisite configuration stuff we're gonna want under the store interface - especially if there's a future where the multisite configuration itself describes what kind of stores get constructed and used. maybe that deserves a separate interface outside of Store? something to think about

**Review by cbodley — APPROVED** [review]

File: src/rgw/rgw_sal.h (hunk 2 above)
Comment (dang) [reply src/rgw/rgw_sal.h:79]: Yeah, that may be.  I'll start thinking about what that would look like.  In the meantime, I'll move these APIs into zipper to help get rid of the ->getRados() calls.

**Comment by dang** [comment]
https://pulpito.ceph.com/dang-2020-09-08_13:50:28-rgw-wip-dang-zipper-8-distro-basic-smithi/

**Pull request merged**